service/apigateway/marshal.go

```diff
@@ -0,0 +1,88 @@
+package apigateway
+
+import (
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/protocol/json"
+	"github.com/aws/aws-sdk-go-v2/aws/protocol/rest"
+)
+
+// protoCreateAPIKeyMarshaler defines a marshaler for CreateApiKey operation
+type protoCreateAPIKeyMarshaler struct {
+	input *CreateApiKeyInput
+}
+
+// marshalOperation is the top level method used within a handler stack to marshal an operation
+// This method calls appropriate marshal shape functions as per the input shape and protocol used by the service.
+func (m protoCreateAPIKeyMarshaler) marshalOperation(r *aws.Request) {
+	var err error
+	encoder := rest.NewEncoder(r.HTTPRequest)
+
+	// We add Content-Type Header if input shape's is not of type streaming payload
+	// The value of Content-Type is decided by following:
+	// a. if shape's metadata has JSONVersion and protocol is json.
+	//    - application/x-amz-json-%s where %s is ths JSONVersion.
+	// b. else if protocol is either json or rest-json
+	//     - application/json
+	// Here protocol is rest-json, and shape is not of type streaming payload,
+	// thus content-type header with value application/json is added.
+	encoder.AddHeader("Content-Type").String("application/json")
+	err = encoder.Encode()
+	if err != nil {
+		r.Error = err
+		return
+	}
+
+	jsonEncoder := json.NewEncoder()
+	err = marshalCreateAPIKeyInputShapeAWSJSON(m.input, r, jsonEncoder)
+	if err != nil {
+		r.Error = err
+	}
+}
+
+// marshalCreateAPIKeyInputShapeAWSJSON is a stand alone function used to marshal the json body
+func marshalCreateAPIKeyInputShapeAWSJSON(v *CreateApiKeyInput, r *aws.Request, enc *json.Encoder) error {
+	if r.Metadata.TargetPrefix != "" {
+		target := r.Metadata.TargetPrefix + "." + r.Operation.Name
+		r.HTTPRequest.Header.Add("X-Amz-Target", target)
+	}
+	if r.Metadata.JSONVersion != "" {
+		jsonVersion := r.Metadata.JSONVersion
+		r.HTTPRequest.Header.Add("Content-Type", "application/x-amz-json-"+jsonVersion)
+	}
+
+	obj := enc.Object()
+	obj.Key("customerId").String(*v.CustomerId)
+	obj.Key("description").String(*v.Description)
+	obj.Key("enabled").Boolean(*v.Enabled)
+	obj.Key("generateDistinctId").Boolean(*v.GenerateDistinctId)
+	obj.Key("name").String(*v.Name)
```

Comment:
dereferencing values should be guarded with a `if v.<Field> != nil`

Reply:
Will add null checks!

Reply:
For marshaling input into JSON, guarding with v.<Field>!=nil might not be enough; If there exists cases where `not setting a key in json` is different than `setting null value for that key`. Those cases, would need customizations.